api/lib/services/csvwBuilder/valueAttribute.ts

```diff
@@ -1,19 +1,63 @@
-import { ResourceIndexer } from '@tpluscode/rdfine'
+import { RdfResource } from '@tpluscode/rdfine'
 import { literal } from '@rdfjs/data-model'
+import { shrink } from '@zazuko/rdf-vocabularies'
 import * as Table from '../../read-model/Table/index'
 import * as Csvw from './index'
-import { csvw } from '../../namespaces'
+import { csvw, rdf } from '../../namespaces'
 
-export function valueAttributeToCsvwColumn (attribute: Table.ValueAttribute, csvwColumn: Csvw.Column & ResourceIndexer) {
-  attribute.parameters && Object.entries(attribute.parameters)
-    .forEach(([param, value]) => {
-      csvwColumn[csvw(param).value] = literal(value)
-    })
+const xsdPrefix = /^xsd:(.+)/
 
+function getBuiltInDatatypeName (datatype: RdfResource): string | null {
+  switch (datatype.id.value) {
+    case (csvw.JSON.value):
+      return 'json'
+    case (rdf.HTML.value):
+      return 'html'
+    case (rdf.XMLLiteral.value):
+      return 'xml'
+  }
+
+  const shrunk = shrink(datatype.id.value)
+  const matches = shrunk.match(xsdPrefix)
+
+  if (!matches) {
+    return null
+  }
+
+  switch (matches[1]) {
+    case 'double':
+      return 'number'
+    case 'base64Binary':
+      return 'binary'
+    case 'dateTime':
+      return 'datetime'
+    case 'anyAtomicType':
+      return 'any'
+  }
+
+  return matches[1]
+}
+
+export function valueAttributeToCsvwColumn (attribute: Table.ValueAttribute, csvwColumn: Csvw.Column) {
   if (attribute.language) {
     csvwColumn.language = attribute.language
   } else if (attribute.datatype) {
-    csvwColumn.datatype = attribute.datatype
+    const builtInType = getBuiltInDatatypeName(attribute.datatype)
+    const hasParameters = attribute.parameters && Object.entries(attribute.parameters).length > 0
+
+    if (builtInType && hasParameters) {
+      const derivedDatatype = csvwColumn._create(csvwColumn._node.blankNode())
+      derivedDatatype[csvw.base.value] = literal(builtInType)
+
+      Object.entries(attribute.parameters)
+        .forEach(([param, value]) => {
+          derivedDatatype[csvw(param).value] = literal(value)
+        })
+
+      csvwColumn.datatype = derivedDatatype
+    } else {
+      csvwColumn.datatype = attribute.datatype
```

Comment:
What ends up in the CSVW if it receives a non-built-in datatype? A type that derives from `string`?

Reply:
no, simply *that* type. any params ignored. although I'm a little confused about the specification. it seems to allow something like

```json
{
  "datatype": {
    "@id": "custom type URI",
    "format": "whatever"
  }
}
```

this does not make sense because if you use the same type for another column, all parameter get mashed together. it only works when strictly following JSON-LD tree structure but once you parse this into RDF it will break

thoughts @ktk @bergos ?

Reply:
As I understand the spec:
```
<csvw column> csvw:datatype xyz:Datatype .
```
This is not valid because it's not a built-in datatype.

```
{
  "datatype": {
    "@id": "custom type URI",
    "format": "whatever"
  }
}
```
And this creates a derived type with `base` = `string` (the default value for `base`)

Reply:
That's not what the spec says IMO. `@id` is allowed  and thus `<> csvw:datatype xyz:Datatype` in turtle is the same as JSON-LD:

```json
{
  "datatype: { 
    "@id": "xyz:Datatype"
  }
}
```

Reply:
Ok that was a misunderstanding on my part. But in this case, it will still always derive from `string`, as defined here: https://www.w3.org/TR/2015/REC-tabular-metadata-20151217/#datatype-base

Reply:
I don't remember the details, but what I concluded in the end:

- If `datatype` is a named node, use it directly
- Otherwise it should be a blank node with `base` which gets expanded with xsd namespace (not 100% sure if that is accurate for all types). Optional there can be a format property. I used it only for date and dateTime datatypes.

Reply:
Yes, that is what we'll provide although it appears that it will not exhaust the csvw spec.